### PR TITLE
MStable

### PR DESCRIPTION
**Pull request by idlehusbandry**
# MStable

Add Mstable strategy

### CODE REVIEW

File: hardhat.config.js [hunk 1]
```diff
@@ -65,8 +65,8 @@ module.exports = {
   networks: {
     hardhat: {
       allowUnlimitedContractSize: true,
-      // forking: {
```
Comment (bugduino) [inline hardhat.config.js:68]: could you please keep the `forking` block commented? in this way unit tests will pass

File: contracts/IdleMStableStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    // _amount is strategy token
+    function redeem(uint256 _amount) external override onlyIdleCDO returns (uint256) {
+        return _redeem(_amount);
+    }
+
+    // _amount in underlying token
+    function redeemUnderlying(uint256 _amount) external override returns (uint256) {
+        uint256 _underlyingAmount = (_amount * oneToken) / price();
+        lastIndexAmount = lastIndexAmount - _underlyingAmount;
```
Comment (bugduino) [inline contracts/IdleMStableStrategy.sol:148]: Why in `redeemUnderlying` the `lastIndexAmount` var is updated while in the `_redeem` it's not? it should be updated in both cases

Comment (bugduino) [reply contracts/IdleMStableStrategy.sol:148]: same for `lastIndexedTime`

File: test/IdleMStableStrategy.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+    await mUSD.connect(user).approve(IdleMStableStrategy.address, AMOUNT_TO_TRANSFER);
+    await IdleMStableStrategy.connect(user).deposit(AMOUNT_TO_TRANSFER);
+
+    let strategySharesAfter = await IdleMStableStrategy.totalSupply();
+
+    let sharesReceived = strategySharesAfter.sub(strategySharesBefore);
+    expect(sharesReceived).gt(0);
+
+    await IdleMStableStrategy.connect(user)["redeemRewards()"]();
```
Comment (bugduino) [inline test/IdleMStableStrategy.js:148]: There are no real tests for this, you should add some assertions

File: test/IdleMStableStrategy.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    await mUSD.connect(musd_signer).transfer(user.address, AMOUNT_TO_TRANSFER);
+    await DAI.connect(dai_signer).transfer(user.address, AMOUNT_TO_TRANSFER);
+
+    await DAI.connect(user).approve(musdSwapingContract.address, AMOUNT_TO_TRANSFER);
+
+    await musdSwapingContract.connect(user).swap(DAIAddress, USDCAddress, AMOUNT_TO_TRANSFER.div(2), 0, user.address);
+
+    await savingsManager.connect(user).collectAndStreamInterest(mUSD.address);
+    
+    console.log("APR", (await IdleMStableStrategy.getApr()).toString());
```
Comment (bugduino) [inline test/IdleMStableStrategy.js:166]: same as above you should add an assertion

**Review by bugduino — CHANGES_REQUESTED** [review]
I added some comments in the PR in general what I think it's missing it's basically, the test for `IdleMStableStrategyWrapper.sol` and more tests for `IdleMStableStrategy.sol` as each single method should be tested with real assertions

File: contracts/IdleMStableStrategy.sol (hunk 2 above)
Comment (idlehusbandry) [reply contracts/IdleMStableStrategy.sol:148]: Changed

File: hardhat.config.js (hunk 1 above)
Comment (idlehusbandry) [reply hardhat.config.js:68]: I using mainnet fork for testing the IdleMstableStrategy. 
If not forked, we would need to deploy MStable, Uniswapv2 locally. 

Any other solution, please suggest.

File: test/IdleMStableStrategy.js (hunk 4 above)
Comment (idlehusbandry) [reply test/IdleMStableStrategy.js:166]: Added the assertion a simple assertion, as it is very difficult to predict the APR.

File: test/IdleMStableStrategy.js (hunk 3 above)
Comment (idlehusbandry) [reply test/IdleMStableStrategy.js:148]: Added assertion to check the raw balance.

File: hardhat.config.js (hunk 1 above)
Comment (bugduino) [reply hardhat.config.js]: you can use it locally no problem on that, just don't commit it so everything will work as outlined here https://github.com/Idle-Labs/idle-tranches#testing

**Comment by bugduino** [comment]
Tests for the strategy wrapper are good. The only thing missing is still some more tests for the main strategy itself (eg missing assertion here https://github.com/Idle-Labs/idle-tranches/pull/19/files#diff-4a3c38e266355ddbdff885aa6a68cf65e7a4a0de253a8f4f89fa9fd1a64ba965R134 and all other methods) and to move the files in the proper folders given that we reorganized the files structure about a month ago for strategies so you should move the mstable files inside a new `mstable` folder in `contracts/`

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        require(_amount != 0, "Deposit amount should be greater than 0");
+        underlyingToken.transferFrom(msg.sender, address(this), _amount);
+        return _depositToVault(_amount);
+    }
+
+    function _depositToVault(uint256 _amount) internal returns (uint256) {
+        underlyingToken.approve(address(imUSD), _amount);
+        lastIndexAmount = lastIndexAmount + _amount;
+        lastIndexedTime = block.timestamp;
+        uint256 interestTokensReceived = imUSD.depositSavings(_amount);
```
Comment (dantop114) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:129]: unused variable!

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            "Should received more reward from uniswap than minLiquidityTokenToReceive"
+        );
+
+        uint256 underlyingBalanceAfter = underlyingToken.balanceOf(address(this));
+        uint256 newCredits = _depositToVault(underlyingBalanceAfter);
+        _mint(msg.sender, newCredits);
+        return newCredits;
+    }
+
+    function claimGovernanceTokens(uint256 startRound, uint256 endRound) public onlyOwner {
```
Comment (dantop114) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:209]: Why is this necessary? Harvests should claim governance tokens when called, this is adding an additional step to this strategy harvesting.

**Review by dantop114 — COMMENTED** [review]
Looks good to me! Just a couple of questions and comments.
In general, good looking strategy. 
The only thing I would encourage is the use of NatSpec.

File: contracts/interfaces/IIdleMStableStrategy.sol [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+import "./IIdleCDOStrategy.sol";
+
+interface IIdleMStableStrategy is IIdleCDOStrategy {
+    function transferShares(
+        address _to,
+        uint256 _interestTokens,
+        uint256 _govShares
+    ) external;
+}
```
Comment (massun-onibakuchi) [inline contracts/interfaces/IIdleMStableStrategy.sol:12]: This interface seems to be unused in `mstable/IdleMstableStrategy.sol`

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    function getRewardTokens() external view override returns (address[] memory) {
+        address[] memory govTokens;
+        govTokens[0] = govToken;
+        return govTokens;
+    }
+
+    function deposit(uint256 _amount) external override onlyIdleCDO returns (uint256 minted) {
+        require(_amount != 0, "Deposit amount should be greater than 0");
```
Comment (massun-onibakuchi) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:120]: If we follow the conventions, we should not revert when `_amount`  is zero.
```
https://github.com/idlehusbandry/idle-tranches/blob/master/contracts/strategies/idle/IdleStrategy.sol#L78

File: contracts/strategies/mstable/IdleMStableStrategy.sol (hunk 6 above)
Comment (idlehusbandry) [reply contracts/strategies/mstable/IdleMStableStrategy.sol:209]: I think `vault.claimRewards()` may take up infinite gas sometimes and may lead to contracts not able to withdraw META tokens. Hence to prevent it I added a function `claimGovernanceTokens(uint256 startRound, uint256 endRound)` which calls `vault.claimRewards(startRound, endRound)`;

Let me know if my thought is wrong, and I will correct the contract accordingly.

File: contracts/interfaces/IIdleMStableStrategy.sol (hunk 7 above)
Comment (idlehusbandry) [reply contracts/interfaces/IIdleMStableStrategy.sol:12]: removed the unused interface

File: contracts/strategies/mstable/IdleMStableStrategy.sol (hunk 8 above)
Comment (idlehusbandry) [reply contracts/strategies/mstable/IdleMStableStrategy.sol]: Yes,
Changed

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        address _underlyingToken,
+        address _vault,
+        address _idleCDO,
+        address _uniswapV2Router02,
+        address[] calldata _routerPath,
+        address _owner
+    ) public initializer {
+        OwnableUpgradeable.__Ownable_init();
+        ReentrancyGuardUpgradeable.__ReentrancyGuard_init();
+
```
Comment (bugduino) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:93]: let's add a require for `token == address(0)` so to avoid initialization of the logic contract (given the `token = address(1);` in the constructor)

File: contracts/strategies/mstable/IdleMStableStrategy.sol (hunk 6 above)
Comment (bugduino) [reply contracts/strategies/mstable/IdleMStableStrategy.sol:263]: what about saving and updating a variable `lastClaimedRound` upon initialize and then on each `claimGovernanceTokens` ?

eg something like this 
```
uint256 endRound = fetchEndRound();
_claimGovernanceTokens(lastClaimedRound, endRound);
lastClaimedRound = endRound;
```
assuming there is a way to fetch the current round number via `fetchEndRound`

File: contracts/strategies/mstable/IdleMStableStrategyWrapper.sol [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import "../../interfaces/IERC20Detailed.sol";
+import "../../interfaces/IMAsset.sol";
+import "../../interfaces/IIdleMStableStrategy.sol";
+
+import "../../IdleCDO.sol";
+
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+
+contract IdleMStableStrategyWrapper {
+    using SafeMath for uint256;
```
[the remark this answers is not in the document]
Comment (bugduino) [reply contracts/strategies/mstable/IdleMStableStrategyWrapper.sol:12]: +1

**Review by bugduino — CHANGES_REQUESTED** [review]

File: contracts/strategies/mstable/IdleMStableStrategyWrapper.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (idlehusbandry) [reply contracts/strategies/mstable/IdleMStableStrategyWrapper.sol:12]: Removed.

File: contracts/strategies/mstable/IdleMStableStrategy.sol (hunk 6 above)
Comment (idlehusbandry) [reply contracts/strategies/mstable/IdleMStableStrategy.sol:263]: The rounds are expected to be calculated offchain and sent as params to `function redeemRewards(bytes calldata _extraData)` 
The function is only callable be idleCDO, hence it this approach.

Comment (bugduino) [reply contracts/strategies/mstable/IdleMStableStrategy.sol:260]: Yes but how? If we can return those value directly in the redeemRewards we can do a static call without any other additional work off-chain basically. If we can keep track of the last claimed round on-chain we probably don't need to pass anything calculated off-chain which will be handy, wdyt?

Comment (idlehusbandry) [reply contracts/strategies/mstable/IdleMStableStrategy.sol:260]: If pass input `(0,0)` to input arguments, the contract will claim all the possible tokens available. The vault contract has functions to solve this.

In short `claimGovernance(0,0)` with withdraw all possible tokens
`claimGoveranance(a,b)` will claim from round a to round b.

https://github.com/idlehusbandry/idle-tranches/blob/2fba9dddab303713cfb0313db268d558eed6e135/contracts/strategies/mstable/IdleMStableStrategy.sol#L268

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 11]
```diff
@@ -123,8 +131,10 @@ contract IdleMStableStrategy is Initializable, OwnableUpgradeable, ERC20Upgradea
     /// @param _extraData must contain the minimum liquidity to receive, start round and end round round for which the reward is being claimed
     /// @return rewards amount of reward that is deposited to vault
     function redeemRewards(bytes calldata _extraData) external override onlyIdleCDO returns (uint256[] memory rewards) {
-        (uint256 minLiquidityTokenToReceive, uint256 startRound, uint256 endRound) = abi.decode(_extraData, (uint256, uint256, uint256));
-        _claimGovernanceTokens(startRound, endRound);
+        (uint256 minLiquidityTokenToReceive, uint256 endRound) = abi.decode(_extraData, (uint256, uint256));
```
Comment (bugduino) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:134]: I think it would be better to avoid passing `endRound` all together and always fetch it via the `unclaimedRewards` call, or do you think that it can take too much gas if not called for a while? Alternatively I think you could add an `if (endRound) == 0)` then fetch the value using `unclaimedRewards` so we have both possibilities available with the 'default' one being the call to unclaimedRewards

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 12]
```diff
@@ -255,24 +265,20 @@ contract IdleMStableStrategy is Initializable, OwnableUpgradeable, ERC20Upgradea
     }
 
     /// @notice Claim governance tokens
-    /// @param startRound Start Round from which the Governance tokens must be claimed
     /// @param endRound End Round from which the Governance tokens must be claimed
-    function claimGovernanceTokens(uint256 startRound, uint256 endRound) public onlyOwner {
-        _claimGovernanceTokens(startRound, endRound);
+    function claimGovernanceTokens(uint256 endRound) public onlyOwner {
```
Comment (bugduino) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:269]: see comment above

File: contracts/strategies/mstable/IdleMStableStrategy.sol [hunk 13]
```diff
@@ -255,24 +265,20 @@ contract IdleMStableStrategy is Initializable, OwnableUpgradeable, ERC20Upgradea
...
+        (, , uint256 _endRound) = vault.unclaimedRewards(address(this));
+        require(endRound > rewardLastRound, "End round should more than the last claimed round ");
```
Comment (bugduino) [inline contracts/strategies/mstable/IdleMStableStrategy.sol:271]: are they doing some of these checks too? what happens if `endRound > _endRound` ?
